Server/src/models/DataUploadModel.ts

```diff
@@ -14,192 +13,697 @@ import { Units } from './entities/Units';
+        this.connection.createQueryBuilder(Publicationtype, 'publicationType')
+            .select('publicationType.id', 'id')
+            .where('LOWER(publicationType.name) = LOWER(:publicationTypeRef)', { publicationTypeRef: publicationTypeReceived })
+            .getRawOne();
+
+    /**
+     * This method will query the DB to see if there is an existing entry, and if so return this
+     * entry. If an entry does not exist, it will return false.
+     * 
+     * @param publicationTypeReceived 
+     * Publication Type: string
+     */
+    private async fetchReferenceTypeId(publicationTypeReceived: string): Promise<any> {
+        let publicationTypeExists = await this.selectReferenceTypeIdQuery(publicationTypeReceived)
+        if (publicationTypeExists == undefined) {
+            publicationTypeExists = false;
+        }
+        return publicationTypeExists;
     }
 
-    async insertPublisher(publisher: string): Promise<number> {
+    /**
+     * This method will create a Publicationtype object and return it's ID. It will check if 
+     * a duplicate Publicationtype exists via @fetchReferenceTypeId and if it exists will set 
+     * the Publicationtype object to have the same ID as the existing entry. If there is no
+     * existing entry, than the method will add a new entry to the Publicationtype table.
+     *
+     * @param publicationTypeReceived 
+     * Publication type: string
+     */
+    async insertReferenceType(publicationTypeReceived: string): Promise<number> {
+        let publicationType = new Publicationtype();
+        publicationType.id;
+        publicationType.name = publicationTypeReceived;
+        let publicationTypeExists: any;
+        publicationTypeExists = await this.fetchReferenceTypeId(publicationTypeReceived);
+        if (publicationTypeExists != false) {
+            publicationType.id = publicationTypeExists.id;
+        }
+        else {
+            await this.connection.manager.save(publicationType);
+        }
+        return publicationType.id;
+    }
 
-        const connection = getConnection();
+    private selectPublisherIdQuery = (publisherReceived: string) =>
+        this.connection.createQueryBuilder(Publisher, 'publisher')
+            .select('publisher.id', 'id')
+            .where('LOWER(publisher.name) = LOWER(:publisherRef)', { publisherRef: publisherReceived })
+            .getRawOne();
+
+    /**
+     * This method will query the DB to see if there is an existing entry, and if so return this
+     * entry. If an entry does not exist, it will return false.
+     * 
+     * @param publisherReceived 
+     * Publisher name: string
+     */
+    private async fetchPublisherId(publisherReceived: string): Promise<any> {
+        let publisherExists = await this.selectPublisherIdQuery(publisherReceived)
+        if (publisherExists == undefined) {
+            publisherExists = false;
+        }
+        return publisherExists;
```

Comment:
This can return either a rowPacket or a boolean. Is there a way we can instead return rowpacket or return nothing instead?

Reply:
Now that you mention it, returning nothing should do the trick, which would allow for cutting a bunch of lines of code. Good catch, I'll see if that works!

Reply:
This was a great thought, it worked perfectly!